collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormsSynchronizerTest.java

```diff
@@ -0,0 +1,128 @@
+package org.odk.collect.android.formmanagement;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.odk.collect.android.forms.Form;
+import org.odk.collect.android.forms.FormRepository;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+import static java.util.Arrays.asList;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@SuppressWarnings("PMD.DoubleBraceInitialization")
+public class ServerFormsSynchronizerTest {
+
+    private ServerFormsSynchronizer synchronizer;
+    private RecordingFormDownloader formDownloader;
+    private FormRepository formRepository;
+    private ServerFormsDetailsFetcher serverFormDetailsFetcher;
+
+    @Before
+    public void setup() {
+        formRepository = new InMemFormRepository();
+        formDownloader = new RecordingFormDownloader();
+        serverFormDetailsFetcher = mock(ServerFormsDetailsFetcher.class);
+
+        synchronizer = new ServerFormsSynchronizer(serverFormDetailsFetcher, formRepository, formDownloader);
+    }
+
+    @Test
+    public void downloadsNewForms() throws Exception {
+        when(serverFormDetailsFetcher.fetchFormDetails()).thenReturn(asList(
+                new ServerFormDetails("form-1", "http://example.com/form-1", null, "form-1", "server", "md5:form-1-hash", null, true, false)
+        ));
+
+        synchronizer.synchronize();
+        assertThat(formDownloader.getDownloadedForms(), containsInAnyOrder("form-1"));
+    }
+
+    @Test
+    public void downloadsUpdatedForms() throws Exception {
+        when(serverFormDetailsFetcher.fetchFormDetails()).thenReturn(asList(
+                new ServerFormDetails("form-1", "http://example.com/form-1", null, "form-1", "server", "md5:form-1-hash", null, false, true)
+        ));
+
+        synchronizer.synchronize();
+        assertThat(formDownloader.getDownloadedForms(), containsInAnyOrder("form-1"));
```

Comment:
Why not cover a little more ground and have a form-1 with some form version in the repository before the synchronization and then assert that the repository contains the form with the new version? That would better match the level of the delete test below, too.

Reply:
Totally agree the test boundaries are inconsistent here. Need to do some refactoring of `MultiFormDownloader` to fix that though. That's something I'd like to fix up in the next PR as I'm going to have to touch the auto download code.